### PR TITLE
Expand multi walker shared resource to ParticleSet and Hamiltonian

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
This PR expands multi walker shared resource beyond TrialWaveFunction to cover ParticleSet and QMCHamiltonian
```
struct FatWalkerResourceCollection
{
  ResourceCollection pset_res;
  ResourceCollection twf_res;
  ResourceCollection ham_res;
}
```
The gold shared resources now are created in the driver. The driver level design is largely complete.

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Review by prckent — CHANGES_REQUESTED** [review]
Some initial comments, both required changes:
1. FatWalker -> ResourcedWalker (or other synonym)
2. The new .h for the above needs a description about what it does, plus function definition comments.

**Comment by ye-luo** [comment]
> Some initial comments, both required changes:
> 
>     1. FatWalker -> ResourcedWalker (or other synonym)

This file only have resource related types right now but it is intended to provide driver based fat walker related types. For example, "WalkerElementsRef.h" will probably be fused into this file.

>     2. The new .h for the above needs a description about what it does, plus function definition comments.

Will do.

**Comment by ye-luo** [comment]
@prckent Rename FatWalker to DriverWalker and add documentation to DriverWalkerTypes.h

**Comment by PDoakORNL** [comment]
I'm reviewing now.

**Review by PDoakORNL — APPROVED** [review]
We have some issues with serialized walkers and resources to work out beyond this scope of this PR.  This looks good as a step forward for shared resource use for the straight forward batched case.

File: src/QMCDrivers/QMCDriverNew.cpp [hunk 1]
```diff
@@ -153,12 +153,21 @@ void QMCDriverNew::startup(xmlNodePtr cur, QMCDriverNew::AdjustedWalkerCounts aw
 
   estimator_manager_->put(population_.get_golden_hamiltonian(), *population_.get_golden_electrons(), cur);
 
+  if (dispatchers_.are_walkers_batched())
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/QMCDriverNew.cpp:156]: This is probably more of my own habit making bool return function names as questions instead of getter.

**Review by prckent — APPROVED** [review]
Thanks for addressing my review comments Ye.

**Pull request merged**